MayssaraOmar/assignment3/Word search/src/Word search.cpp

```diff
@@ -2,47 +2,54 @@
 #include "word search.h"
 using namespace std;
 
-void get_words_from_grid(vector<vector<char> > &grid, int row, int col,
-		string word, map<pair<int, int>, bool> &vis,
-		map<pair<pair<int, int>, string>, bool> &memo, dictionary &dic,
-		unordered_set<string> &words) {
+// internal function
+// cache maps a pair of a (cell position (row, col)) and a string (prefix) to a boolean to ensure that (get_words_from_grid) method doesn't run for the same inputs more than once using memoization technique
+static void get_words_from_grid(const vector<vector<char> > &grid, int row, int col,
+		string prefix, map<pair<int, int>, bool>* visited,
+		map<pair<pair<int, int>, string>, bool>* cache, const dictionary &dic,
+		unordered_set<string>* words) {
```

Comment:
Some comments on the function / parameters:
I like the cache, avoiding checking twice the same thing, but why do you think is important? How likely it's that you will hit it? 
I think the visited parameter will be slightly better (more performant) as a vector<vector<bool>>

Reply:
I think it depends on the type of inputs, if there is a lot of duplicates nearby in the grid, then the cache might be useful to optimize time. However, in all other types of input, it reserves unnecessary memory.